balancer-js/src/modules/zaps/bbausd2-migrations/stables.ts

```diff
@@ -0,0 +1,250 @@
+   * @param {{string, string, string}}  to Pool info being migrated to
+   * @param {string}                    bptIn Amount of BPT tokens to migrate.
+   * @param {string}                    minBptOut Minimum of expected BPT out ot the migration flow.
+   * @param {boolean}                   staked Indicates whether tokens are initially staked or not.
+   * @param {string[]}                  underlyingTokens Underlying token addresses. Array must have the same length and order as underlying tokens in pool being migrated from. Refer to [getPoolTokens](https://github.com/balancer-labs/balancer-v2-monorepo/blob/master/pkg/interfaces/contracts/vault/IVault.sol#L334).
+   * @param {string}                    authorisation Encoded authorisation call.
+   * @returns Migration transaction request ready to send with signer.sendTransaction
+   */
+  calldata(
+    userAddress: string,
+    from: { id: string; address: string; gauge?: string },
+    to: { id: string; address: string; gauge?: string },
+    bptIn: string,
+    minBptOut: string,
+    staked: boolean,
+    underlyingTokens: string[],
+    authorisation?: string
+  ): {
+    to: string;
+    data: string;
+  } {
+    const relayer = this.addresses.relayer;
+    let calls: string[] = authorisation
+      ? [this.buildSetRelayerApproval(authorisation)]
+      : [];
+
+    if (staked && (from.gauge == undefined || to.gauge == undefined))
+      throw new Error(
+        'Staked flow migration requires gauge addresses to be provided'
+      );
+
+    if (staked) {
+      calls = [
+        ...calls,
+        this.buildWithdraw(userAddress, bptIn, from.gauge as string),
+        this.buildExit(from.id, relayer, bptIn, underlyingTokens),
+        this.buildSwap(minBptOut, relayer, to.id, to.address, underlyingTokens),
+        this.buildDeposit(userAddress, to.gauge as string),
+      ];
+    } else {
+      calls = [
+        ...calls,
+        this.buildExit(from.id, userAddress, bptIn, underlyingTokens),
+        this.buildSwap(
+          minBptOut,
+          userAddress,
+          to.id,
+          to.address,
+          underlyingTokens
+        ),
+      ];
+    }
+
+    const callData = balancerRelayerInterface.encodeFunctionData('multicall', [
+      calls,
+    ]);
+
+    return {
+      to: this.addresses.relayer,
+      data: callData,
+    };
+  }
+
+  /**
+   * Encodes exitPool call data.
+   * Exit stable pool proportionally to underlying stables. Exits to relayer.
+   * Outputreferences are used to store exit amounts for next transaction.
+   *
+   * @param {string}    poolId Pool id.
+   * @param {string}    sender Sender address.
+   * @param {string}    amount Amount of BPT to exit with.
+   * @param {string[]}  underlyingTokens Token addresses to exit to.
+   * @returns Encoded exitPool call. Output references.
+   */
+  buildExit(
+    poolId: string,
+    sender: string,
+    amount: string,
+    underlyingTokens: string[]
```

Comment:
are underlyingTokens = poolTokens?

Reply:
Underlying tokens is normally the name used (afaik) for the main tokens in a boostedPool, e.g. DAI/USDC/USDT. But in this case I think poolTokens works too as they are the tokens of the source stable pool. I don't think assets is a great name as it doesn't really link it to a pool in any way. Happy to go with the majority.

Reply:
I agree that assets isn't a great name.
I believe that using poolTokens will depend on wether we expect it to include the BPT for that pool or not. Querying for poolTokens on bbausd2 will return bb-a-USD within the token array. This is not the case for stabal3. How do these stable pools compare to that? I ask because this parameter must exclude the BPT for that pool. 
Anyway, I'd say underlyingTokens seems less ambiguous.

Reply:
Assets is what what smart contracts are using for swaps / joins / exits.
Seems like one name has a different meaning in different contexts:

poolTokens = tokens registered in a pool
aka assets = poolTokens in swap / join / exit functions
poolToken = BPT, "balancer pool token" aka liquidity pool token

I'm not feeling entirely comfortable with adding `underlying` as yet another one to the mix, that's why suggested assets to follow contracts for consistency.

If we really need a new name then it would be great to start some kind of naming guideline somewhere.